apis/python/src/tiledbsoma/io/spatial/_util.py

```diff
@@ -37,39 +75,249 @@ def _version_less_than(version: str, upper_bound: Tuple[int, int, int]) -> bool:
+
+        # Var values.
+        self._var_name: npt.NDArray | None = None  # features/name
+        self._gene_id: npt.NDArray | None = None  # features/id (unique identifier)
+        self._feature_type: npt.NDArray | None = None  # features/feature_type
+        self._genome: npt.NDArray | None = None  # features/genome
+
+    def __enter__(self) -> Self:
+        self.open()
+        return self
+
+    def __exit__(self, *_: Any) -> None:
+        self.close()
+
+    def _read_software_version(self) -> tuple[int, int, int]:
+        if self._root is None:
+            raise RuntimeError(f"[internal] File '{self._path}' is not open reading.")
         try:
-            version = dataset.attrs["software_version"]
+            raw_version = self._root.attrs["software_version"]
         except KeyError as ke:
             raise SOMAError(
                 f"Unable to read software version from gene expression file "
-                f"{gene_expression_path}."
+                f"{self._path}."
             ) from ke
-    if not isinstance(version, str):
-        raise SOMAError(
-            f"Unexpected type {type(version)!r} for software version in gene "
-            f"expression file {gene_expression_path}. Expected a string."
-        )
-    version_prefix = "spaceranger-"
-    if not version.startswith(version_prefix):
-        raise SOMAError(
-            f"Unexpected value {version} for software version in gene expresion "
-            f"file {gene_expression_path}."
-        )
-    version = version[len(version_prefix) :].split(".")
-    if len(version) not in {3, 4}:
-        raise SOMAError(
-            f"Unexpected value {version} for software version in gene expresion "
-            f"file {gene_expression_path}."
-        )
-    try:
-        major = _str_to_int(version[0])
-        minor = _str_to_int(version[1])
-        patch = _str_to_int(version[2])
-    except ValueError:
-        raise SOMAError(
-            f"Unexpected value {version} for software version in gene expresion "
-            f"file {gene_expression_path}."
-        )
-    return (major, minor, patch)
+        if not isinstance(raw_version, str):
+            raise SOMAError(
+                f"Unexpected type {type(raw_version)!r} for software version in gene "
+                f"expression file {self._path}. Expected a string."
+            )
+        version: str | list[str] = raw_version.split("-")
+        if len(version) == 1:
+            version = version[0]
+        elif len(version) == 2:
+            self._software = version[0]
+            version = version[1]
+        else:
+            raise SOMAError(
+                f"Unexpected value {raw_version} for 'software_version' in gene "
+                f"expression file {self._path}."
+            )
+
+        version = version.split(".")
+        if len(version) not in {3, 4}:
+            raise SOMAError(
+                f"Unexpected value {raw_version} for 'software_version' in gene "
```

Comment:
Maybe give examples here in the error string of what acceptable formats would be